articles/WideStrings.md

```diff
@@ -0,0 +1,186 @@
+UTF-8 and UTF-16 are both variable width encodings, meaning a code point can take 1 to 4 bytes depending on the encoding.
+It may take multiple code points to represent a single user perceived character.
+One of the goals of ROS 2 is to support microcontrollers that are constrained by both code size and processor speed.
+Some wide string operations like splitting a string on a user perceived character may not be possible on these devices.
+
+However, whole string equality checking is the same whether using wide strings or not.
+Further splitting a UTF-8 string on an ASCII character is identical to splitting an ASCII character on an ASCII string.
+If code on a microcontroller must do string manipluation then it could assert that a `string` only contains ASCII data by ceasing to proces a string when it encounters a byte greater than 127.
+
+## What does the API look like to a user of ROS 2?
+
+### Python 3
+
+In Python the `str` type will be used for both strings and wide strings.
+Bytes of a known encoding should be converted to a `str` using [bytes.decode](https://docs.python.org/3/library/stdtypes.html#bytes.decode) before being assigned to a field.
+
+**Example**
+
+```python
+import sys
+from time import sleep
+
+import rclpy
+from rclpy.qos import qos_profile_default
+from std_msgs.msg import WString
+
+
+if __name__ == '__main__':
+    rclpy.init(sys.argv)
+
+    node = rclpy.create_node('talker')
+
+    chatter_pub = node.create_publisher(WString, 'chatter', qos_profile_default)
+
+    msg = WString()
+    msg.data = 'Hello World' + bytes([0x00, 0x00, 0x26, 0x3A]).decode('utf-32-be')
+    print('Publishing: "{0}"'.format(msg.data))
+    chatter_pub.publish(msg)
+    node.destroy_node()
+    rclpy.shutdown()
+```
+
+### C++
+
+In c++ wsing `wchar_t` has different sizes on different platforms (2 bytes on Windows, 4 bytes on Linux).
+Instead ROS 2 will use `char16_t` for characters of wide strings, and `std::u16string` for wide strings themselves.
+
+**Example**
+
+```
+#include <codecvt>
+#include <iostream>
+#include <memory>
+
+#include "rclcpp/rclcpp.hpp"
+
+#include "std_msgs/msg/wstring.hpp"
+
+int main(int argc, char * argv[])
+{
+  rclcpp::init(argc, argv);
+
+  auto node = rclcpp::node::Node::make_shared("talker");
+
+  rmw_qos_profile_t custom_qos_profile = rmw_qos_profile_default;
+  custom_qos_profile.depth = 7;
+
+  auto chatter_pub = node->create_publisher<std_msgs::msg::WString>("chatter", custom_qos_profile);
+
+  rclcpp::WallRate loop_rate(2);
+
+  auto msg = std::make_shared<std_msgs::msg::WString>();
+  auto i = 1;
+  std::wstring_convert<std::codecvt<char16_t, char, std::mbstate_t>, char16_t> convert_to_u16;
+
+  while (rclcpp::ok()) {
+    std::u16string hello(u"Hello World: " + convert_to_u16.from_bytes(std::to_string(i++)));
+    msg->data = hello;
+    std::cout << "Publishing: '" << msg->data.cpp_str() << "'" << std::endl;
```

Comment:
Use `printf` according to our guidelines.

Reply:
`std::printf` in 5f3c97fd8326deb728fe3ebf6f573212bd8d2646